csrc/multidevice/utils.cpp

```diff
@@ -151,6 +217,116 @@ void insertReshardings(Fusion* fusion) {
   }
 }
 
+void insertShardedAxisReordering(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  std::vector<Expr*> reshard_exprs;
+  for (auto expr : exprs) {
+    if (isResharding(expr)) {
+      reshard_exprs.push_back(expr);
+    }
+  }
+  for (auto expr : reshard_exprs) {
+    NVF_ERROR(
+        ir_utils::isTvOp(expr),
+        "Non-tv op is not supported yet: ",
+        expr->toString());
+    NVF_ERROR(
+        expr->outputs().size() == 1,
+        "Resharding operations can only have one output");
+    NVF_ERROR(
+        expr->inputs().size() == 1,
+        "Resharding operations can have only one input");
+    auto output = expr->outputs().at(0)->as<TensorView>();
+    auto input = expr->inputs().at(0)->as<TensorView>();
+    auto [shard_additions, shard_deletions] = getShardingChanges(expr);
+    NVF_ERROR(
+        shard_additions.size() + shard_deletions.size() <= 1,
+        "Resharding expr can only support one axis")
+
+    // For gather operations i.e. ID goes from sharded to unsharded
+    // this will rematerialize a sharded axis.
+    // ProcessGroup expects contiguous tensors.
+    // Update input to push the rematerialized axis to the front -> collective
+    // -> permute the rematerizlied axis to the proper location Example: [i0
+    // DIDx(i1)] -> [i0 i1] Rewritten to: [DIDx(i1) i0] -> [i1 i0] -> [i0 i1]
+    // Note: there are no reduction based collectives that materializes an axis
+    // so expr is guaranteed to be a set.
+    if (!shard_deletions.empty()) {
+      IterDomain* shard_deleted_id = shard_deletions[0];
+      int sharding_axis =
+          static_cast<int>(input->domain()->rootPosOf(shard_deleted_id));
+      if (isOutermostAllocatedId(input, shard_deleted_id)) {
+        continue;
+      }
+      TensorView* input_permute = permute(input, {{sharding_axis, 0}});
```

Comment:
This permute is moving a DeviceDim axis to position 0, so it's actually guaranteed to be DeviceDim. We know that in the input this axis must be DeviceDim and in the output it must be Serial since the expression is unsharding that axis.